internal/runners/revert/revert.go

```diff
@@ -0,0 +1,107 @@
+package revert
+
+import (
+	"strconv"
+
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/internal/primer"
+	"github.com/ActiveState/cli/internal/prompt"
+	"github.com/ActiveState/cli/pkg/cmdlets/commit"
+	"github.com/ActiveState/cli/pkg/platform/api/mono/mono_models"
+	"github.com/ActiveState/cli/pkg/platform/model"
+	"github.com/ActiveState/cli/pkg/project"
+	"github.com/go-openapi/strfmt"
+)
+
+type Revert struct {
+	out     output.Outputer
+	prompt  prompt.Prompter
+	project *project.Project
+}
+
+type Params struct {
+	CommitID string
+}
+
+type primeable interface {
+	primer.Outputer
+	primer.Prompter
+	primer.Projecter
+}
+
+func New(prime primeable) *Revert {
+	return &Revert{
+		prime.Output(),
+		prime.Prompt(),
+		prime.Project(),
+	}
+}
+
+type commitDetails struct {
+	Date        string
+	Author      string
+	Description string
+	Changeset   []changeset `locale:"changeset,Changes"`
+}
+
+type changeset struct {
+	Operation   string `locale:"operation,Operation"`
+	Requirement string `locale:"requirement,Requirement"`
+}
+
+func (r *Revert) Run(params *Params) error {
+	commitID := strfmt.UUID(params.CommitID)
+	revertCommit, err := model.GetCommit(strfmt.UUID(params.CommitID))
+	if err != nil {
+		return locale.WrapError(err, "err_revert_get_commit", "Could not fetch commit details for commit with ID: {{.V0}}", params.CommitID)
+	}
+
+	history, fail := model.CommitHistory(r.project.Owner(), r.project.Name())
+	if fail != nil {
+		return locale.WrapError(fail.ToError(), "err_revert_get_history", "Could not get project commit history")
+	}
+	if !containsCommitID(history, commitID) {
+		return locale.NewError("err_revert_invalid_commit_id", "Commit ID: {{.V0}} does not belong to the project: {{.V1}}", params.CommitID, r.project.Namespace().String())
+	}
+
+	count, fail := model.CommitsBehindLatest(r.project.Owner(), r.project.Name(), r.project.CommitID())
+	if fail != nil {
+		return locale.WrapError(fail.ToError(), "err_revert_commits_behind", "Could not determine if local project is synchronized with platform")
+	}
```

Comment:
Just wondering:  Do we actually need to run all these checks?  I would expect them to happen on the backend already.  This, of course, makes it easier to write out helpful and localized messages, and before asking the user to confirm. So, I am not opposed to the choice.  I am just wondering.  Is the dashboard doing the same?

Reply:
Yes the backend will return an error in both cases I've covered above. I thought this would be a good way to provide more contextual errors but I think it's doing too much. I've updated the error handling. If a user gives a commit ID that covers the two cases above (and possibly more) they will receive:

```
Could not revert to commit: <commitID> please ensure that the local project is synchronized with the platform and that the given commit ID belongs to the current project
Error occurred while trying to update a branch: Commit parent doesn't match branch commit
```

This is a fairly long error string but provides some remedies.